File: Tree/leetcode_99.py
# Problem Link ====>>https://leetcode.com/problems/recover-binary-search-tree/description/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        def inorder(node, prev, first, second):
            if not node:
                return

            # Traverse left subtree
            inorder(node.left, prev, first, second)

            # Check for swapped nodes
            if prev and node.val < prev.val:
                if not first:
                    first = prev
                second = node
            prev = node

            # Traverse right subtree
            inorder(node.right, prev, first, second)

        prev = None
        first, second = None, None
        inorder(root, prev, first, second)

        # Swap the values of the two nodes
        first.val, second.val = second.val, first.val


# Example usage
root = TreeNode(3)
root.left = TreeNode(1)
root.right = TreeNode(4)
root.right.left = TreeNode(2)

sol = Solution()
sol.recoverTree(root)

# Print the corrected BST (in-order traversal)


def print_inorder(node):
    if not node:
        return
    print_inorder(node.left)
    print(node.val, end=' ')
    print_inorder(node.right)


print_inorder(root)

File: Tree/leetcode_99.cpp
// Problem Link ====>>https://leetcode.com/problems/recover-binary-search-tree/description/

#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
public:
    void recoverTree(TreeNode *root)
    {
        TreeNode *first = nullptr;  // First swapped node
        TreeNode *second = nullptr; // Second swapped node
        TreeNode *prev = nullptr;   // Previous node during in-order traversal

        // In-order traversal to find swapped nodes
        inorder(root, prev, first, second);

        // Swap the values of the two nodes
        swap(first->val, second->val);
    }

private:
    void inorder(TreeNode *root, TreeNode *&prev, TreeNode *&first, TreeNode *&second)
    {
        if (!root)
            return;

        // Traverse left subtree
        inorder(root->left, prev, first, second);

        // Check for swapped nodes
        if (prev && root->val < prev->val)
        {
            if (!first)
                first = prev;
            second = root;
        }
        prev = root;

        // Traverse right subtree
        inorder(root->right, prev, first, second);
    }
};

int main()
{
    // Example usage
    TreeNode *root = new TreeNode(3);
    root->left = new TreeNode(1);
    root->right = new TreeNode(4);
    root->right->left = new TreeNode(2);

    Solution sol;
    sol.recoverTree(root);

    // Print the corrected BST (in-order traversal)
    // You can implement the printInorder function similar to the previous examples
    // For simplicity, I'm not printing the result here

    return 0;
}
